L1Trigger/L1TTrackMatch/interface/L1TrackJetEmulationProducer.h

```diff
@@ -7,73 +7,356 @@
-  static const float Z0_LSB_POW = pow(2.0, -1 * Z0_EXTRABITS);
-  static const float Z0_LSB = 0.05 * Z0_LSB_POW;
-  inline float floatPt(pt_intern pt) { return pt.to_float(); }
-  inline int intPt(pt_intern pt) { return (ap_ufixed<18 + PT_EXTRABITS, 13 + PT_EXTRABITS>(pt)).to_int(); }
-  inline float floatEta(glbeta_intern eta) { return eta.to_float() * ETA_LSB; }
-  inline float floatPhi(glbphi_intern phi) { return phi.to_float() * PHI_LSB; }
-  inline float floatZ0(z0_intern z0) { return z0.to_float() * Z0_LSB; }
-
-  inline pt_intern makePt(int pt) { return ap_ufixed<18 + PT_EXTRABITS, 13 + PT_EXTRABITS>(pt); }
-  inline pt_intern makePtFromFloat(float pt) { return pt_intern(INTPT_LSB_POW * round(pt / INTPT_LSB_POW)); }
-  inline z0_intern makeZ0(float z0) { return z0_intern(round(z0 / Z0_LSB)); }
-
-  inline ap_uint<pt_intern::width> ptToInt(pt_intern pt) {
-    // note: this can be synthethized, e.g. when pT is used as intex in a LUT
-    ap_uint<pt_intern::width> ret = 0;
-    ret(pt_intern::width - 1, 0) = pt(pt_intern::width - 1, 0);
-    return ret;
-  }
+const unsigned int PT_INTPART_BITS{9};
+const unsigned int ETA_INTPART_BITS{3};
+const unsigned int kExtraGlobalPhiBit{4};
 
-  inline glbeta_intern makeGlbEta(float eta) { return round(eta / ETA_LSB); }
-  inline glbeta_intern makeGlbEtaRoundEven(float eta) {
-    glbeta_intern ghweta = round(eta / ETA_LSB);
-    return (ghweta % 2) ? glbeta_intern(ghweta + 1) : ghweta;
-  }
+typedef ap_ufixed<TTTrack_TrackWord::TrackBitWidths::kRinvSize - 1, PT_INTPART_BITS, AP_TRN, AP_SAT> pt_intern;
+typedef ap_fixed<TTTrack_TrackWord::TrackBitWidths::kTanlSize, ETA_INTPART_BITS, AP_TRN, AP_SAT> glbeta_intern;
+typedef ap_int<TTTrack_TrackWord::TrackBitWidths::kPhiSize + kExtraGlobalPhiBit> glbphi_intern;
+typedef ap_int<TTTrack_TrackWord::TrackBitWidths::kZ0Size> z0_intern;  // 40cm / 0.1
+typedef ap_uint<TTTrack_TrackWord::TrackBitWidths::kD0Size> d0_intern;
 
-  inline glbphi_intern makeGlbPhi(float phi) { return round(phi / PHI_LSB); }
+inline const unsigned int DoubleToBit(double value, unsigned int maxBits, double step) {
+  unsigned int digitized_value = std::floor(std::abs(value) / step);
+  unsigned int digitized_maximum = (1 << (maxBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+  if (digitized_value > digitized_maximum)
+    digitized_value = digitized_maximum;
+  if (value < 0)
+    digitized_value = (1 << maxBits) - digitized_value;  // two's complement encoding
+  return digitized_value;
+}
+inline const double BitToDouble(unsigned int bits, unsigned int maxBits, double step) {
+  int isign = 1;
+  unsigned int digitized_maximum = (1 << maxBits) - 1;
+  if (bits & (1 << (maxBits - 1))) {  // check the sign
+    isign = -1;
+    bits = (1 << (maxBits + 1)) - bits;
+  }
+  return (double(bits & digitized_maximum) + 0.5) * step * isign;
+}
 
-};  // namespace convert
+// eta/phi clusters - simulation
+struct EtaPhiBin {
```

Comment:
Many of the names in this header sounds general-enough to have risk of name collisions in the global namespace. E.g. `EtaPhiBin` is defined already here
https://github.com/cms-sw/cmssw/blob/f65893802e7b24f5460f6e314f682e5a8268d965/L1Trigger/L1TTrackMatch/interface/L1Clustering.h#L10-L20
and different from the definition here, leading to violating the One Definition Rule. It would be better to enclose the definitions of such types and functions into a namespace.

Furthermore, if a header is (intended to be) used only in the code in `plugins` (or only in `src`), it would be better to move the header to `plugins` (or `src`), because it is an implementation detail that does not need to be part of the public interface of the package. If the header is (intended to be) used only from one source file, moving the contents into the source file would allow to use anonymous namespace to easily avoid name collisions.

Reply:
I updated the PR to comply with both. Thanks